ethereum/core/src/main/java/org/hyperledger/besu/ethereum/trie/diffbased/bonsai/storage/flat/BonsaiArchiveFlatDbStrategy.java

```diff
@@ -0,0 +1,451 @@
+import org.hyperledger.besu.ethereum.trie.diffbased.common.storage.flat.CodeStorageStrategy;
+import org.hyperledger.besu.metrics.BesuMetricCategory;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+import org.hyperledger.besu.plugin.services.metrics.Counter;
+import org.hyperledger.besu.plugin.services.storage.SegmentedKeyValueStorage;
+import org.hyperledger.besu.plugin.services.storage.SegmentedKeyValueStorageTransaction;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+
+import kotlin.Pair;
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.bouncycastle.util.Arrays;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class BonsaiArchiveFlatDbStrategy extends BonsaiFullFlatDbStrategy {
+  private static final Logger LOG = LoggerFactory.getLogger(BonsaiArchiveFlatDbStrategy.class);
+
+  protected final Counter getAccountFromArchiveCounter;
+  protected final Counter getStorageFromArchiveCounter;
+
+  public BonsaiArchiveFlatDbStrategy(
+      final MetricsSystem metricsSystem, final CodeStorageStrategy codeStorageStrategy) {
+    super(metricsSystem, codeStorageStrategy);
+
+    getAccountFromArchiveCounter =
+        metricsSystem.createCounter(
+            BesuMetricCategory.BLOCKCHAIN,
+            "get_account_from_archive_counter",
+            "Total number of calls to get account that were from archived state");
+
+    getStorageFromArchiveCounter =
+        metricsSystem.createCounter(
+            BesuMetricCategory.BLOCKCHAIN,
+            "get_storage_from_archive_counter",
+            "Total number of calls to get storage that were from archived state");
+  }
+
+  static final byte[] MAX_BLOCK_SUFFIX = Bytes.ofUnsignedLong(Long.MAX_VALUE).toArrayUnsafe();
+  static final byte[] MIN_BLOCK_SUFFIX = Bytes.ofUnsignedLong(0L).toArrayUnsafe();
+  public static final byte[] DELETED_ACCOUNT_VALUE = new byte[0];
+  public static final byte[] DELETED_CODE_VALUE = new byte[0];
+  public static final byte[] DELETED_STORAGE_VALUE = new byte[0];
+
+  private Optional<BonsaiContext> getStateArchiveContextForWrite(
+      final SegmentedKeyValueStorage storage) {
+    // For Bonsai archive get the flat DB context to use for writing archive entries. We add one
+    // because
+    // we're working with the latest world state so putting new flat DB keys requires us to +1 to it
+    Optional<byte[]> archiveContext = storage.get(TRIE_BRANCH_STORAGE, WORLD_BLOCK_NUMBER_KEY);
+    if (archiveContext.isPresent()) {
+      try {
+        return Optional.of(
+            // The context for flat-DB PUTs is the block number recorded in the specified world
+            // state, + 1
+            new BonsaiContext(
+                Long.decode("0x" + (new String(archiveContext.get(), StandardCharsets.UTF_8)))
+                    + 1));
```

Comment:
I think you can use something like that to decode and encode in the database

encode to byte[] before writing into the database
    byte[] bytes = Bytes.ofUnsignedLong(100000).toArrayUnsafe();

decode from the database
Bytes.wrap(bytes.toArrayUnsafe()).toLong();

Reply:
How we manage a reorg in this case if we add everytime +1 ? maybe I'm missing something

Reply:
> I think you can use something like that to decode and encode in the database
> 
> encode to byte[] before writing into the database byte[] bytes = Bytes.ofUnsignedLong(100000).toArrayUnsafe();
> 
> decode from the database Bytes.wrap(bytes.toArrayUnsafe()).toLong();

Yeah I'll look at rewriting that in the couple of places it's used, thanks